modules/core/src/main/java/org/apache/fluo/core/impl/TransactionImpl.java

```diff
@@ -905,19 +905,21 @@ public int getSize() {
   }
 
   private <V> void addCallback(CompletableFuture<V> cfuture, CommitData cd,
-      OnSuccessInterface onSuccessInterface) {
-    cfuture.thenAcceptAsync(result -> {
-      try {
-        // this method currently throws Exception, so must be caught...
-        onSuccessInterface.onSuccess(cd, result);
-      } catch (Exception e) {
-        cd.commitObserver.failed(e);
+      OnSuccessInterface<V> onSuccessInterface) {
+    cfuture.handleAsync((result, exception) -> {
+      if (result != null) {
+        try {
+          onSuccessInterface.onSuccess(cd, result);
+          return null;
+        } catch (Exception e) {
+          cd.commitObserver.failed(e);
+          return null;
+        }
+      } else {
```

Comment:
If a future completes successfully and returns null, then do not want to take the exception handling path.  Maybe the code should do :

```java
 if(exception != null) {
   //handle exception
 } else {
    //handle result
 }
```